fix: handle DOS-style path for Dockerfile in minikube build

ComradeProgrammer opened the pull request:
<!-- 🎉 Thank you for contributing to minikube! 🎉 Here are some hints to get your PR merged faster:

1. Your PR title will be included in the release notes, choose it carefully
2. If the PR fixes an issue, add "fixes #<issue number>" to the description.
3. If the PR is a user interface change, please include a "before" and "after" example.
4. If the PR is a large design change, please include an enhancement proposal:
https://github.com/kubernetes/minikube/tree/master/enhancements
-->
fixes #15007

medyagh commented on cmd/minikube/cmd/image.go:
```diff
@@ -310,6 +310,8 @@ var buildImageCmd = &cobra.Command{
 			}
 		}
+		//if dockerFile is a DOS path, translate it into UNIX path
+		dockerFile = strings.ReplaceAll(dockerFile, "\\", "/")
```
@ComradeProgrammer do you mind pasting an example Dockerfile and the error that you get?

this change seems to be dangrous, I suggest to limit this line to only run on Windows !
it is better that the user Fixes the Dockerfile themselves and minikube gives them an educational instruciton how to do that

ComradeProgrammer replied:
Well, I don't think it is a problem which can be solved by user's dockerfile. It is normal to include Dockerfile in the .dockerignore so that the Dockerfile itself won't be packed into the image.

However, when user is using minikube on windows, in order to prevent the Dockerfile from being ignored when we try to tar the workspace, user need to specify the path to the Dockerfile, which should be a DOS-style path. However, when this image is built, the image is built on the virtual machine we created, which requires a UNIX-style path.

Therefore if user pass the DOS-style path of Dockerfile, virtual machine cannot build this image because the path is not  a valid UNIX path; if user pass the UNIX-style path, the Dockerfile will be ignored and wont't be packed into the tar file which we send to the virtual machine.

Unless users waive the thought of preventing Dockerfile to be packed into image, and remove Dockerfile from .dockerignore, there is no way for user to prevent this bug from being triggered.

ComradeProgrammer replied:
The example has been updated into the comment of this pr. Besides, changes have been made to limit this line to only run on Windows

ComradeProgrammer commented:
Here is an example to reproduce the problem. The folder is
```
.
├── .dockerignore
└── test
    └── Dockerfile
```

The content of .dockerignore
```
**/Dockerfile*
```

The content of dockerfile
```
FROM ubuntu:latest

CMD ["/bin/bash", "-c"]
```

Output:
When I execute ` .\minikube-windows-amd64.exe image build -t aaa:latest -f test\Dockerfile .`, using DOS-style path
```
PS D:\workspace\minikube> .\minikube-windows-amd64.exe image build -t aaa:latest -f test\Dockerfile .
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat var/lib/minikube/build/build.519194499/test\Dockerfile: no such file or directory
```
When I execute ` .\minikube-windows-amd64.exe image build -t aaa:latest -f test\Dockerfile .`, using UNIX-style path
```
PS D:\workspace\minikube> .\minikube-windows-amd64.exe image build -t aaa:latest -f test/Dockerfile .
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /var/lib/minikube/build/build.4139476469/test/Dockerfile: no such file or directory
```

spowelljr commented on cmd/minikube/cmd/image.go:
```diff
@@ -310,6 +310,10 @@ var buildImageCmd = &cobra.Command{
 			}
 		}
+		if runtime.GOOS == "windows" {
+			//if dockerFile is a DOS path, translate it into UNIX path
```
```suggestion
			// if dockerFile is a DOS path, translate it into UNIX path
```

spowelljr commented:
Tested this and confirms it works

**Before:**
```
> minikube image build -t thing:latest -f thing2\Dockerfile .
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /var/lib/minikube/build/build.3965132197/thing2\Dockerfile: no such file or directory
```

**After:**
```
> minikube image build -t thing:latest -f thing2\Dockerfile .
Sending build context to Docker daemon   2.56kB
Step 1/2 : FROM ubuntu:latest
 ---> 216c552ea5ba
Step 2/2 : CMD ["/bin/bash", "-c"]
 ---> Running in 5b64e4107d63
Removing intermediate container 5b64e4107d63
 ---> a5b667236e51
Successfully built a5b667236e51
Successfully tagged thing:latest
```

medyagh reviewed (changes requested):
@ComradeProgrammer Lets make sure this replace is only suggested by minikube and not automatically done, if minikube gonna replace it, it has to get user permission to clearly say "Yes" it is better to leave this operation to the user

ComradeProgrammer commented:
> @ComradeProgrammer Lets make sure this replace is only suggested by minikube and not automatically done, if minikube gonna replace it, it has to get user permission to clearly say "Yes" it is better to leave this operation to the user

This function has been added, now the we need user's permission to replace the \

spowelljr commented on cmd/minikube/cmd/image.go:
```diff
@@ -310,6 +312,22 @@ var buildImageCmd = &cobra.Command{
+			out.String("minikube detects that you are using DOS-style path %s. Should minikube convert it to UNIX-style by replacing all \\ to /? [y/n]", dockerFile)
+			var answer string
+			for {
+				fmt.Scan(&answer)
```
Should check for error

spowelljr replied:
Many people use minikube in their CI for testing. If someone is running this command in a script it will hang forever since there will be no input.

Either some sort of interaction flag will need to be added so users running via scripts can bypass this, or instead of automating this we can just suggest that the user makes this change themself.

spowelljr commented on cmd/minikube/cmd/image.go:
```diff
@@ -310,6 +312,22 @@ var buildImageCmd = &cobra.Command{
+					break
+				}
+			}
+			if answer == "y" || answer == "yes" {
```
`answer` is always going to be empty here because you're creating `answer` with a new scope within the `for` loop on L322 because of the `:=`. If the `:` is removed it will update the value instead of creating a new one.

ComradeProgrammer commented:
@spowelljr Indeed you are right, so I changed the code. Instead of asking for user's input, I switched to command line arguments.
I added a param called "--no-dospath-conversion". By default the DOS-like path is automatically converted, but if user use this --no-dospath-conversion then there will be no conversion. Do you think it is better this time?

spowelljr reviewed (changes requested):
`make generate-docs` to update the new command on the website

ComradeProgrammer commented:
Anything I need to do next?
@medyagh  @spowelljr  @afbjorklund

andez2000 commented:
Is this going to get done?  I have an issue in my build with specifying the files to the command line

spowelljr commented on test/integration/functional_test.go:
```diff
@@ -299,9 +299,10 @@ func validateImageCommands(ctx context.Context, t *testing.T, profile string) {
-
-		if _, err := Run(t, exec.CommandContext(ctx, Target(), "-p", profile, "ssh", "pgrep", "buildkitd")); err == nil {
-			t.Errorf("buildkitd process is running, should not be running until `minikube image build` is ran")
+		if runtime.GOOS != "windows" {
```
I think there might have been some confusion on my last comment. We should check do this check on Windows as well, so the if statement should be removed.

What I was trying to say is that on line 311, if not on Windows the path to the Dockerfile should continue being `filepath.Join(*testdataDir, "build")`, but on Windows we should use `"testdata\\build"` to test this change.

ComradeProgrammer replied:
Actually I found that on Windows, the result of filepath.Join(*testdataDir, "build") is exactly "testdata\\build", so actually we don't need to modify this part, or write a seperate test.

spowelljr replied:
Ah, gotcha, thanks for the info

spowelljr commented on test/integration/functional_test.go:
```diff
@@ -299,7 +299,7 @@ func validateImageCommands(ctx context.Context, t *testing.T, profile string) {
 	t.Run("ImageBuild", func(t *testing.T) {
 		MaybeParallel(t)
-
+		
```
Could you remove this whitespace change?

ComradeProgrammer commented:
@spowelljr  It seems that the test under Docker_Windows has passed

spowelljr commented on cmd/minikube/cmd/image.go:
```diff
@@ -310,6 +311,12 @@ var buildImageCmd = &cobra.Command{
 			}
 		}
+		if runtime.GOOS == "windows" && strings.Contains(dockerFile, "\\") {
+			// if dockerFile is a DOS path, and user did not explicitly forbid this conversion, translate it into UNIX path
```
```suggestion
			// if dockerFile is a DOS path, translate it into UNIX path
```
With the changes the user can no longer forbid this, so comment is incorrect

spowelljr reviewed (approved):
PR looks good, thanks for all the changes
